### PR TITLE
Defer LockFileItem.Properties dictionary construction until needed

### LINKED ISSUE #13557: PERF: Unnecessary construction of LockFileItem.Properties dictionary
### NuGet Product Used

Other/NA

### Product Version

VS17

### Worked before?

N/A

### Impact

Other

### Repro Steps & Context

See https://github.com/NuGet/NuGet.Client/pull/5861 for more details

Open VS
Wait for things to calm down
Start profiling
Open Roslyn.sln, open a file it none opened
Wait for things to calm down
Stop profiling

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by ToddGrun**
## Bug
-- 

Fixes: https://github.com/NuGet/Home/issues/13557

-- Helps improve memory allocations during solution load

Regression? Last working version:
-- N/A

## Description

-- Defers creation of a dictionary until it's needed. From the testing I've done on the Roslyn sln, nearly half of these dictionaries need not be created.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] Test exception
-- Changes are trivial enough that tests wouldn't be useful
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by ToddGrun** [comment]
Repro:

Open VS
Wait for things to calm down
Start profiling
Open Roslyn.sln, open a file it none opened
Wait for things to calm down
Stop profiling

Small improvement: ~15MB less allocated during sln load

Allocations during Utf8JsonStreamLockFileItemConverter:

*** Previous ***
image

*** With changes ***
image

File: src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs [hunk 1]
```diff
@@ -18,7 +18,19 @@ public LockFileItem(string path)
...
 
-        public IDictionary<string, string> Properties { get; } = new Dictionary<string, string>();
+        private Dictionary<string, string> _properties;
+        public IDictionary<string, string> Properties
+        {
+            get
+            {
+                if (_properties == null)
+                {
+                    _properties = new Dictionary<string, string>();
```
Comment (ToddGrun) [inline src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs]: DIdn't know what the threading semantics are around this. Should I go ahead and put in a lock such that multiple calls into the same object can't possible both create a dictionary?

**Comment by ToddGrun** [comment]
Is the failing test (MainInternal_OnParsingSuccess_DoesNotShowsHelp) flaky? It passes for me locally.

File: src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs]: In practice, the construction/reading are single threaded. 
The safer approach might be a lock though.

Comment (ToddGrun) [reply src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs]: Went ahead and put this behind a static lock

**Review by nkolev92 — APPROVED** [review]

**Comment by ToddGrun** [comment]
@nkolev92 , @Nigusu-Allehu -- Do tests need kicked or should I dig a bit more?

**Comment by nkolev92** [comment]
I retriggered the tests.

btw, the fixes statement needs to have the link right after it so that GH autocloses. 
It's the GitHub way of closing issues automatically when a PR merges.

**Comment by ToddGrun** [comment]
> I retriggered the tests.

Looks like some mac tests failed on the rerun. Is this something I should look into more?

**Comment by nkolev92** [comment]
> > I retriggered the tests.
> 
> Looks like some mac tests failed on the rerun. Is this something I should look into more?

No need. I think this is ready.

**Pull request merged**